### PR TITLE
[ML] Correct ML exceptions to use RestStatus

### PR DESCRIPTION
**Pull request by maxhniebergall**
Serverless has forced us to reassess our exceptions. Specifically, we are replacing instances of ElasticsearchException with ElasticsearchStatusException, so that only exceptions which require Elastic support are thrown as ElasticsearchExceptions or with 500 (InternalServerError) status codes. 

I greped for ElasticsearchException in the ML code base and replaced nearly all instances with the status exception. Elasticians can view my explanations [here](https://docs.google.com/document/d/1oiVoQ8HtLLzgzHM6is2Zeux1b7WchMU9mMCzprJaDIA/edit?usp=sharing).

Still remaining after this PR are the [serverError functions in the ExceptionsHelper class](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/ExceptionsHelper.java#L61-L73). These changes were not included in this PR because they appear to be quite extensive, and this PR is already fairly large.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java:27]: This logic can work on `ElasticsearchException` rather than `ElasticsearchStatusException`, because `ElasticsearchException` has the `status` method.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java:40]:
```suggestion
                } else if (firstCode != elasticsearchStatusException.status().getStatus()) {
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java:57]: I think for this case it's better to return a code that actually happened rather than always return timeout.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java:196]: This could be `TOO_MANY_REQUESTS` as it's really saying "retry later".

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/ResultsPersisterService.java:238]: This could be `TOO_MANY_REQUESTS` as it's really saying "retry later".

**Review by droberts195 — COMMENTED** [review]
Thanks for doing this. I've left quite a few comments but this is still mostly on the right track.

> Still remaining after this PR are the [serverError functions in the ExceptionsHelper class](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/ExceptionsHelper.java#L61-L73).

I think in the cases where we've chosen to call `serverError` it's likely reasonable to really report a server error. So let's leave those as-is for now.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteForecastAction.java
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteForecastAction.java:256]: You can use `instanceof` pattern matching to avoid the later cast

```suggestion
            if (e instanceof ElasticsearchException elasticsearchException) {
                listener.onFailure(
                    new ElasticsearchStatusException(
                        "An error occurred while searching forecasts to delete",
                        elasticsearchException.status(),
```

Comment (maxhniebergall) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportDeleteForecastAction.java:256]: awesome!

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredResultsRemover.java
Comment (maxhniebergall) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/retention/ExpiredResultsRemover.java:123]: I think we should still be returning an elasticsearchStatusException here, no?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java
Comment (maxhniebergall) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/ExceptionCollectionHandling.java:57]: I removed the branch for having the same status. Now, this function returns the first exception if they are all 4XX statuses, and 500 status if any of the statuses are outside of the 4xx range or are not elasticsearchExceptions.

**Review by davidkyle — APPROVED** [review]
LGTM

**Review by droberts195 — APPROVED** [review]
LGTM if you could just incorporate my last few minor comments before merging.

**Comment by maxhniebergall** [comment]
@elasticsearchmachine run elasticsearch-ci/docs

**Pull request merged**
